definition_validator/validator.cc

```diff
@@ -271,7 +271,7 @@ void validateCompatibleOverride(const core::Context ctx, core::MethodRef superMe
             }
 
             // if there is a corresponding parameter, make sure it has the right type
-            if (corresponding != right.kw.required.end() && corresponding != right.kw.optional.end()) {
+            if (corresponding != right.kw.required.end() || corresponding != right.kw.optional.end()) {
```

Comment:
Thanks for the find!

I think we want to go one step further. It's never ok to compare one iterator with another iterator as the vectors might be incompletely different places.

So I think that in the case where, on line 269, we overwrite `corresponding` to hold an iterator for `optional`, we can no longer compare `corresponding` against `right.kw.required.end` (because in that case, `corresponding` is holding an iterator ver `right.kw.optional`).

I believe you in saying that the `||` is enough to fix the problem, but I would love if we could restructure this code to hoist the comparison `==` comparison into a local variable, and then use that in both places.

```cpp
            auto hasCorrespondingRequired = corresponding != right.kw.required.end()
            if (!hasCorrespondingRequired) {
                corresponding =
                    absl::c_find_if(right.kw.optional, [&](const auto &r) { return r.get().name == req.get().name; });
            }

            auto hasCorrespondingOptional = corresponding != right.kw.optional.end();
            // if there is a corresponding parameter, make sure it has the right type
            if (hasCorrespondingRequired || hasCorrespondingOptional) {
```

Reply:
Good point, I used local variables instead 👍